Optimize dep-graph serialization and improve meta-data hashing

nikomatsakis opened the pull request:
This makes a number of related changes:

- instead of serializing the whole graph, we only serialize `HIR -> WorkProduct` edges
- we gather predecessors more efficiently and cache effort, drastically cutting serialization time by over 6x in my measurements
- compute the metadata hashes more determinstically
- improve `-Z incremental-info`, which tells you how much re-use you are getting, and tells you *why* you are not getting re-use

The end result is that, at least in my tests, we are now able to get 100% re-use when doing `cargo clean; cargo build` twice in a row (with an incremental directory, of course). Here are some rough performance numbers from the `syntex-syntax` crate:

|  build | normal time (debug) | incremental time (debug) |
| ----- | ---- | -------- |
|  full build | 128s | 78s |
|  just syntex_syntax | 96s | 51s |

Note that these numbers only reflect the first phase of incremental, where we are just skipping the trans/LLVM phases. They will get better as we incrementalize more of the compiler. Also, these numbers represent the simplest case, where no actual changes were made. There is still more work to do to handle the case some small number of changes were made correctly (in particular, verifying that the dep-graph is complete etc).

r? @michaelwoerister 

Fixes #35167.

michaelwoerister commented:
Are these numbers for a release or for a debug build?

michaelwoerister commented on src/librustc_incremental/persist/hash.rs:
```diff
@@ -39,6 +39,14 @@ impl<'a, 'tcx> HashContext<'a, 'tcx> {
         }
     }
 
+    pub fn is_hashable(dep_node: &DepNode<DefId>) -> bool {
```
I'd prefer if this was called `is_input_node` or `is_graph_root` to make it clearer what we are after.

michaelwoerister commented on src/librustc_incremental/persist/load.rs:
```diff
@@ -110,157 +110,106 @@ pub fn decode_dep_graph<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    // Compute the set of Hir nodes whose data has changed or which have been removed.
+    let dirty_raw_source_nodes = dirty_nodes(tcx, &serialized_dep_graph.hashes, &retraced);
+
+    // Create a (maybe smaller) list of
```
This comment seems unfinished.

michaelwoerister commented on src/librustc_incremental/persist/save.rs:
```diff
@@ -98,38 +101,37 @@ fn save_in<F>(sess: &Session, opt_path_buf: Option<PathBuf>, encode: F)
+    let mut edges = vec![];
+    for (&target, sources) in &preds.inputs {
+        match *target {
+            DepNode::MetaData(_) => continue, // see encode_metadata_hashes instead
```
Wait, isn't this ignoring MetaData items that are inputs? Doesn't encode_metadata_hashes only deal with exported MetaData items?

nikomatsakis replied:
@michaelwoerister 

> Wait, isn't this ignoring MetaData items that are inputs? Doesn't encode_metadata_hashes only deal with exported MetaData items?

Actually, we only expect local-data here -- I should add an assert. Remember that this is matching just on the *target* of an edge, and we have reduce the set of edges to direct hops between an input (hir or foreign metadata) to an output (workproduct or local metadata). Should be ok, iow, but I agree confusing and could use an assert.

nikomatsakis replied:
Arguably I should add a new `DepNode` for "local metadata" vs "foreign metadata".

michaelwoerister replied:
Oh right, it's just matching on the target. Nevermind then.
Regarding a separate `DepNode` variant: That would make sense. Also, with some additional machinery, similar to what we do for object files, we could integrate the metadata-object into the dependency graph too and avoid rewriting it if nothing has changed (which it shouldn't need to if there are only changes in unexported functions).

nikomatsakis replied:
> Also, with some additional machinery, similar to what we do for object files, we could integrate the metadata-object into the dependency graph too 

Yes, I've thought about this. We have basically all the data we need, I think, but we have to be able to save/reload the meta-data. One problem with this is that it includes def-ids which may have changed and would have to be retraced. (This is why we don't just hash the raw bytes of the metadata.) I guess we could encode meta-data using the same "def-path-index" system, so that we do the work of retracing when we load the meta-data, instead of when we write it. In short, we have to think about this a bit, but it's all definitely doable.

michaelwoerister commented on src/librustc_incremental/persist/save.rs:
```diff
@@ -141,70 +143,70 @@ pub fn encode_dep_graph<'a, 'tcx>(hcx: &mut HashContext<'a, 'tcx>,
+        // Create a vector containing a pair of (source-id, hash).
+        // The source-id is stored as a `DepNode<u64>`, where the u64
+        // is the det. hash of the def-path. This is convenient
+        // because we can sort this to get a table ordering across
```
"table" -> "stable"

michaelwoerister commented on src/librustc_incremental/persist/save.rs:
```diff
@@ -141,70 +143,70 @@ pub fn encode_dep_graph<'a, 'tcx>(hcx: &mut HashContext<'a, 'tcx>,
+        // The source-id is stored as a `DepNode<u64>`, where the u64
+        // is the det. hash of the def-path. This is convenient
+        // because we can sort this to get a table ordering across
+        // compilations, even if the def-ids themselves have changed.
```
smart :)

michaelwoerister commented:
Ok, I've reviewed everything after `rustfmt save.rs`. The question in `encode_dep_graph` seems important.

nikomatsakis commented:
@michaelwoerister OK, I addressed your nits. I will rebase atop of #35166, I think, and maybe try to land them together once you've revised the additional commits in there.

michaelwoerister commented on src/librustc_incremental/persist/load.rs:
```diff
@@ -113,10 +113,28 @@ pub fn decode_dep_graph<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    // encoding, rather than having been retraced to a `DefId`. The
+    // reason for this is that this way we can include nodes that have
+    // been removed (which no longer have a `DefId` in the current
+    // compilation).
```
:+1:

michaelwoerister commented on src/librustc_incremental/persist/load.rs:
```diff
@@ -113,10 +113,28 @@ pub fn decode_dep_graph<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    // retracted-target-node) edges. In the process of retracing the
+    // target nodes, we may discover some of them def-paths no longer exist,
+    // in which case there is no need to mark the corresopnding nodes as dirty
+    // (they are just not present). So this list may be smaller than the original.
```
:+1:

michaelwoerister commented on src/librustc_incremental/persist/save.rs:
```diff
@@ -110,7 +110,12 @@ pub fn encode_dep_graph(preds: &Predecessors,
+                // those in `encode_metadata_hashes`, which comes later.
+                assert!(def_id.is_local());
+                continue;
+            }
```
:+1:

michaelwoerister commented:
LGTM.
